### PR TITLE
feat: hook support via useSynapse

### PR DESCRIPTION
**Pull request by kylealwyn**
## Summary
WIP PR to introduce a hook helper into synaptik. If you haven't already, give [hooks](https://reactjs.org/docs/hooks-intro.html) a look.

## Todo
- [x] Settle on name (useSynaptik, useVault, useConnect, useStores)
- [x] Write [tests](https://reactjs.org/blog/2019/02/06/react-v16.8.0.html#testing-hooks)

### CODE REVIEW

File: src/synaptik.js [hunk 1]
```diff
@@ -177,3 +176,48 @@ export function connect(selector, lifecycle = {}) {
...
 }
+
+export const useSynaptik = selector => {
+  const vault = useContext(Context);
+  const [state, setState] = useState(selector(vault.stores));
+
+  useEffect(() => {
+    return vault.subscribe(() => {
+      const nextState = selector(vault.stores);
+      if (!isShallowEqual(nextState, state)) {
```
Comment (cehsu) [inline src/synaptik.js]: Will we ever want to check for deep equality?

**Review by cehsu — COMMENTED** [review]
Thanks for the sneak peak of this exciting work 👶

File: src/synaptik.js (hunk 1 above)
Comment (kylealwyn) [reply src/synaptik.js]: I don't think so. I'd like to stick to the functionality of the current `PureComponent` implementation which only does the shallow reference and equality check. Deep equality checks come at a performance cost that we don't necessarily need to pay for

File: src/Connect.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    let { lifecycle, synapse } = this.props;
+
+    this.unsubscribe = synapse.subscribe(() => {
+      this.setState(this.getObservedState());
+    });
+
+    if (lifecycle.didMount) lifecycle.didMount(...this.getArgs());
+  }
+
+  componentDidUpdate() {
```
Comment (erikshestopal) [inline src/Connect.js:32]: Any interest in passing down `oldProps`  to `lifecycle.didUpdate`?

Comment (kylealwyn) [reply src/Connect.js:32]: @erikshestopal I'd like to deprecate Connect/connect as we move towards hook adoption. I haven't seen in either form of connect recently - was there a use case you needed oldprops?

File: README.md [hunk 3]
```diff
@@ -149,6 +151,38 @@ export default class TodoList extends Component {
...
+    todos: stores.todos.state.entries,
+    input: stores.todos.state.input,
+    updateInput: stores.todos.updateInput,
+    addTodo: stores.todos.addTodo,
+  }));
+
+  return (
+    <>
+      <ul>
+        {todos.map(todo => (
```
Comment (choochootrain) [inline README.md]: `state.todos`?

File: src/useSynapse.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import React, { useContext, useState, useEffect } from 'react';
+import Context from './Context';
+import shallowEqual from './shallow-equal';
+
+export default function useSynapse(selector) {
+  const synapse = useContext(Context);
+  const select = () => selector(synapse.stores, synapse.state);
```
Comment (choochootrain) [inline src/useSynapse.js]: do we need to expose both `synapse.stores` and `synapse.state`?

Comment (kylealwyn) [reply src/useSynapse.js]: nah

**Comment by kylealwyn** [comment]
@sappira-inc/engineering any final feedback here?

**Review by choochootrain — APPROVED** [review]
lgtm

**Pull request merged**